[WIP] RD Training (#57521)
odoo/odoo: doc/howtos/rdtraining/compute_onchange.rst

:ref:`reference/fields/compute`.

.. note::

    **Goal**: at the end of this section:

    - On the property model, the total area and the best offer are computed:

    .. image:: compute_onchange/media/compute.gif
        :align: center
        :alt: Compute fields

    - On the property offer model, the validity date is computed and has an inverse:

    .. image:: compute_onchange/media/compute_inverse.gif
        :align: center
        :alt: Compute field with inverse

In our real estate module, we have defined the living area as well as the garden area. It is then
natural to define the total area as the sum of both fields. We will use the concept of computed
field for this: the value of a given field is computed from the value of other fields.

So far fields have been stored directly in and retrieved directly from the
database. Fields can also be *computed*. In that case, the field's value is not
retrieved from the database but computed on-the-fly by calling a method of the
model.

To create a computed field, create a field and set its attribute
:attr:`~odoo.fields.Field.compute` to the name of a method. The computation

> method should simply set the value of the field to compute on every record in

Reviewer: We can precise that the convention for the method is to start with '_compute' (as it's done in the previous chapter for the fields finishing with _id or _ids)

    - Add the ``total_area`` field to ``estate.property``. It is defined as the sum of the
      ``living_area`` and the ``garden_area``.

    - Add the field in the form view as depicted on the first image of the **Goal**.

Relational fields can also be used as dependencies::

    description = fields.Char(compute="_compute_description")
    partner_id = fields.Many2one("res.partner")

    @api.depends("partner_id.name")
    def _compute_description(self):
        for record in self:
            record.description = "Test for partner %s" % record.partner_id.name

The example is given with a :class:`~odoo.fields.Many2one`, but it is valid for
:class:`~odoo.fields.Many2many` or a :class:`~odoo.fields.One2many`. An example can be found
`here <https://github.com/odoo/odoo/blob/713dd3777ca0ce9d121d5162a3d63de3237509f4/addons/account/models/account_reconcile_model.py#L248-L251>`__.

Let's try it on our module with the following exercise!

.. exercise:: Compute the best offer

    - Add the ``best_price`` field to ``estate.property``. It is defined as the maximum of the
      offers' ``price``.

    - Add the field in the form view as depicted on the first image of the **Goal**.

>     Tip: you might give a try to the :meth:`~odoo.models.BaseModel.mapped` method. See

Reviewer 2: ```suggestion
    Tip: you might want to give a try to the :meth:`~odoo.models.BaseModel.mapped` method. See
```

        _name = "test.computed"

        total = fields.Float(compute="_compute_total", inverse="_inverse_total")
        amount = fields.Float()

        @api.depends("amount")
        def _compute_total(self):
            for record in self:
                record.total = 2.0 * record.amount
        
        def _inverse_total(self):
            for record in self:
                record.amount = record.total / 2.0

An example can be found
`here <https://github.com/odoo/odoo/blob/2ccf0bd0dcb2e232ee894f07f24fdc26c51835f7/addons/crm/models/crm_lead.py#L308-L317>`__.

.. exercise:: Compute a validity date for offers

    - Add the following fields to the ``estate.property.offer`` model:

    ========================= ========================= =========================
    Field                     Type                      Default
    ========================= ========================= =========================
    validity                  Integer                   7
    date_deadline             Date
    ========================= ========================= =========================

    The ``date_deadline`` is a computed field defined as the offer ``create_date`` to which the

>     ``validity`` is added. Define the appropriate inverse function so that the user can set the

Reviewer 2: ```rst
The ``date_deadline`` is a computed field which is defined as the addition of two fields from the offer: the ``create_date`` and the ``validity``.
```

    date or the validity.

    Tip: the ``create_date`` is only filled in when the record is created. At creation, you will
    need a fallback to prevent crashing.

>     - Add the fields in the form and list view as depicted on the second image of the **Goal**.

Reviewer 2: Even if it's shown in the video, I think it's easy to miss the fact that we have to save the property in order for the change to be applied. It's quite different from the behavior of `compute`.

Author: note added

.. exercise:: Set values to garden area and orientation

    Create an ``onchange`` on the ``estate.property`` model in order to give a value to the
    garden area (10) and orientation (North) when the garden is set. When unset, clean the fields.

Additional Information
----------------------

Onchanges methods can also return a non-blocking warning message
(`example <https://github.com/odoo/odoo/blob/cd9af815ba591935cda367d33a1d090f248dd18d/addons/payment_authorize/models/payment.py#L34-L36>`__).

How to use them?
================

There is no strict rule on the use of computed fields and onchanges.

In many cases, both computed fields and onchanges may be used to achieve the same result. Always
prefer computed fields since they are also triggered outside of the context of a form view. Never
ever use an onchange to add business logic to your model. This is a **very bad** idea since
onchanges are not automatically triggered when creating a record programmatically; they are only
triggered on the form view.

The usual pitfall of computed fields and onchanges is trying to be 'too smart' by adding too much

> logic. This can have the opposite result than expected: the end user is confused about all the

Reviewer 2: ```suggestion
logic. This can have the opposite result of what you expected: the end user is confused about all the
```

automation.

Computed fields tend to be easier to debug: such a field is set by a given method, so it's easy to

> track when the value is set. Onchanges, on the other, may be confusing: it is very difficult to

Reviewer 2: ```suggestion
track when the value is set. Onchanges, on the other hand, may be confusing: it is very difficult to
```

``compute`` method is called at each change of the dependencies.

.. exercise:: Compute a validity date for offers

    - Add the following fields to the ``estate.property.offer`` model:

    ========================= ========================= =========================
    Field                     Type                      Default
    ========================= ========================= =========================
    validity                  Integer                   7
    date_deadline             Date
    ========================= ========================= =========================

    The ``date_deadline`` is a computed field which is defined as the addition of two fields from
    the offer: the ``create_date`` and the ``validity``. Define the appropriate inverse function
    so that the user can set the date or the validity.

    Tip: the ``create_date`` is only filled in when the record is created. At creation, you will
    need a fallback to prevent crashing.

    - Add the fields in the form and list view as depicted on the second image of the **Goal**.

Additional Information
----------------------

> Computed fields are **not stored** in the database. A side-effect is that it is **not

Reviewer 3: ```suggestion
By default, computed fields are **not stored** in the database. A side-effect is that it is **not
```
